### PR TITLE
[fx] Speedup ShapeEnv cache invalidation checks

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112687
* #112933

This may seem a bit silly but we spend ~5% of compilation on simply checking if the `ShapeEnv` cache has been invalidated. It isn't necessarily slow, but we call it millions of times per compile so everything adds up.

To improve the situation, I've added a version counter to the shape env that gets incremented whenever the cache key changes. This does require a bit of care in `ShapeEnv` that we don't modify the relevant state without calling `self._update_version_counter()`. However, we already have a similar situation for the translation validation feature which requires `_set_replacement` to be called instead of modifying the replacements directly.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -3114,7 +3138,7 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
             try:
                 r = try_solve(expr, mod_expr, floordiv_inequality=False)
                 if r is not None and r[1] == 0:
-                    self.divisible.add(mod_expr)
+                    self._add_divisible(mod_expr)
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:3141]: This is used just here, perhaps just inline it?

**Review by lezcano — COMMENTED** [review]
This looks good to me, but it is true that makes the whole system a bit more difficult to use, because if you don't update the cache manually you get a tricky bug. At the very least you should update the `_lru_cache` docs reflecting this.

Technically, the PR looks correct to me, but I'll let @ezyang assess whether we are fine with this optimisation.

**Review by ezyang — APPROVED** [review]
This is fine but it would be nice to have some debug mode where we check `self._get_key` anyway and error if it's inconsistent with the version counter, similar to how we run Z3 translation validation on everything.

**Pull request closed without merging**